### PR TITLE
Pulseaudio: Add function to handle Pulseaudio pa_operation

### PR DESCRIPTION
**Pull request by illuusio**
This PR unifies handling on pa_operation. It does not add or remove anything just adds function that should make easier not to get into infinite loop at the start.

### CODE REVIEW

**Comment by RossBencina** [comment]
There does not appear to be a reason for this to be a macro, could you make it a regular function please?

**Comment by illuusio** [comment]
Ok I've reworked macro to function and made needed changes to code.

File: src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c [hunk 1]
```diff
@@ -105,6 +105,34 @@ int PaPulseAudio_updateTimeInfo( pa_stream * s,
     return 0;
 }
 
+/* Release pa_operation always same way */
+int PaPulseAudio_ReleaseOperation(PaPulseAudio_HostApiRepresentation *hostapi,
```
Comment (RossBencina) [inline src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c:109]: I'm not sure whether you should be returning an error if there is a timeout, but right now the code does not return a value so I suggest...

```suggestion
void PaPulseAudio_ReleaseOperation(PaPulseAudio_HostApiRepresentation *hostapi,
```

File: src/hostapi/pulseaudio/pa_linux_pulseaudio_internal.h [hunk 2]
```diff
@@ -195,11 +195,14 @@ PaPulseAudio_Stream;
...
             } \
     } \
-    if( !pastream->isActive || pastream->isStopped ) \
+    if( !(pastream)->isActive || (pastream)->isStopped ) \
     { \
-            return paStreamIsStopped; \
+        return paStreamIsStopped; \
     }
 
+int PaPulseAudio_ReleaseOperation(PaPulseAudio_HostApiRepresentation *hostapi,
```
Comment (RossBencina) [inline src/hostapi/pulseaudio/pa_linux_pulseaudio_internal.h:203]:
```suggestion
void PaPulseAudio_ReleaseOperation(PaPulseAudio_HostApiRepresentation *hostapi,
```

File: src/hostapi/pulseaudio/pa_linux_pulseaudio.c [hunk 3]
```diff
@@ -1461,11 +1448,8 @@ PaError PaPulseAudio_RenameSource( PaStream *s, const char *streamName )
     op = pa_stream_set_name( stream->inputStream, streamName, RenameStreamCb, stream );
     PaPulseAudio_UnLock( stream->mainloop );
 
-    /* Wait for completion. */
-    while (pa_operation_get_state( op ) == PA_OPERATION_RUNNING)
-    {
-        pa_threaded_mainloop_wait( stream->mainloop );
-    }
+    PaPulseAudio_ReleaseOperation( stream->hostapi,
+                                   &op );
```
Comment (philburk) [inline src/hostapi/pulseaudio/pa_linux_pulseaudio.c:1452]: The code that this replaces did not call pa_operation_unref().
Does this new code fix an old problem with missing unref calls?

File: src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c [hunk 4]
```diff
@@ -105,6 +105,34 @@ int PaPulseAudio_updateTimeInfo( pa_stream * s,
...
+    pa_operation *localOperation = (*operation);
+
+    while( pa_operation_get_state( localOperation ) == PA_OPERATION_RUNNING )
+    {
+        pa_threaded_mainloop_wait( hostapi->mainloop );
+
+        wait ++;
+        usleep( 1000 );
+
+        if( wait > 2000 )
```
Comment (RossBencina) [inline src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c:122]: this condition should be checked at the top of the block, immediately after the while condition indicates still running

**Review by RossBencina — CHANGES_REQUESTED** [review]

File: src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c [hunk 5]
```diff
@@ -105,6 +105,34 @@ int PaPulseAudio_updateTimeInfo( pa_stream * s,
...
+/* Release pa_operation always same way */
+int PaPulseAudio_ReleaseOperation(PaPulseAudio_HostApiRepresentation *hostapi,
+                                  pa_operation **operation)
+{
+    unsigned int wait = 0;
+    pa_operation *localOperation = (*operation);
+
+    while( pa_operation_get_state( localOperation ) == PA_OPERATION_RUNNING )
+    {
+        pa_threaded_mainloop_wait( hostapi->mainloop );
```
Comment (philburk) [inline src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c:117]: According to:
https://freedesktop.org/software/pulseaudio/doxygen/thread-mainloop_8h.html#ac96d19260567d0406350f1b3a14c4bed

the call to pa_threaded_mainloop_wait() needs to be under a lock.

File: src/hostapi/pulseaudio/pa_linux_pulseaudio.c (hunk 3 above)
Comment (illuusio) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio.c:1452]: `pa_operation_unref` should be called after process is completed. Code was incorrect.

File: src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c (hunk 5 above)
Comment (illuusio) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c:117]: Ok made needed changes to have locking when using wait

File: src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c (hunk 4 above)
Comment (illuusio) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c:122]: Moved to correct place

**Comment by illuusio** [comment]
@philburk this could be reviewed again is it ok to merge

File: src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c (hunk 5 above)
Comment (philburk) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c:117]: I cannot find the commits where the locks were added.

File: src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c (hunk 4 above)
Comment (RossBencina) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c:122]: I don't see your change. Did you forget to push it?

Comment (illuusio) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c:123]: My bad! Not pushed stuff as you mentioned. Now done it and rebased.

File: src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c (hunk 5 above)
Comment (illuusio) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c:118]: My bad! Not pushed stuff as you mentioned. Now done it and rebased.

**Comment by illuusio** [comment]
It's seem that this need bit rework as I you reopen device releasing operation won't work as expected.

File: src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c [hunk 6]
```diff
@@ -897,6 +911,7 @@ static PaError RequestStop( PaPulseAudio_Stream * stream,
     PaError ret = paNoError;
     PaPulseAudio_HostApiRepresentation *pulseaudioHostApi = stream->hostapi;
     pa_operation *pulseaudioOperation = NULL;
+    int waitLoop = 0;
```
Comment (RossBencina) [inline src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c:914]: The `waitLoop` variable seems unused.

Comment (RossBencina) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c:914]: @illuusio please fix this before we merge.

**Review by philburk — APPROVED** [review]

File: src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c (hunk 6 above)
Comment (illuusio) [reply src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c]: Sorry this took too long but now it's fixed.

**Review by RossBencina — APPROVED** [review]

**Review by philburk — APPROVED** [review]

**Pull request merged**